### PR TITLE
Proposal: Build-time and runtime detection for hardware acceleration with advance CPU features

### PR DESCRIPTION
**Pull request by yuhaoth**
## Description

When implement #5387, we found `illegal instruction` . With a global compiler flag, we can not resolve that for runtime detection case. The issue is reported at #5758, I think that can be resolved together.

And there are 5 hardware acceleration that implement their own runtime detection. 

Beside that,  we also need hardware acceleration only configuration for each algorithm to reduce code size. Some of algorithms can be implemented with two different hardware accelerations, we need some solution to cover that.

## Gatekeeper checklist

- [ ] **changelog** provided, or not required
- [ ] **backport** done, or not required
- [ ] **tests** provided, or not required

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

### CODE REVIEW

**Comment by yuhaoth** [comment]
@gilles-peskine-arm , I am preparing a demo PR. That base on #6918. And I think we should add some new compilers for improve test coverage.

- aarch64-none-elf , Aarch64 baremetal target
- arm-none-linux-gnueabihf, Aarch32 linux target
- arm-none-eabi, upgrade to support Armv8 A32 and T32 test.
Above compilers can be download from https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads

Add qemu-static for improving test coverage. It can be download from https://github.com/multiarch/qemu-user-static/releases . It can be used to test runtime detection. For timebeing , I think `qemu-aarch64`,`qemu-arm`,`qemu-i386`, `qemu-x84_64` should be put into test image.  Then we can control the cpu feature sets to improve test coverage of hardware specification code.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
The general idea looks fine, but some of the specifics are unclear or problematic.

I found the general structure of the document hard to follow: it's not always clear what is a mandatory requirement (and why), what is a logical consequence, and what is a weighted decision. I recommend the following structure for an architecture document:

1. Short introduction. The one here is ok.
2. Definition of requirements. The requirements here aren't very good because they already assume some of the solution (e.g. “runtime detection for architecture must be enabled” — what does enabled mean here?).
3. Explore the possible solutions. Clearly distinguish between things that _must_ be a certain way, and places where there are several possible choices. Where there are multiple choices, explain why you're making one particular choice .
4. Present the chosen solution. In this section, refer to the exploration section for justification, but don't present alternatives.
5. Any remaining open questions. Those are things that don't need to be decided before the beginning of the project, but will be decided by the time the project is complete (or sometimes will be deferred to a future version).

**Comment by yuhaoth** [comment]
Thanks your suggestions, I will followup it. 

> but some of the specifics are unclear or problematic.

Yes. I am not very sure some points when I wrote it.  #7078 includes my test codes.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+```c
+#if defined(MBEDTLS_ARCH_IS_ARM64)
+
+#if defined(MBEDTLS_RUNTIME_DETECTION)
+#if defined(__clang__)
+#   if __clang_major__ < 4
+#       error "A more recent Clang is required for MBEDTLS_AES_C"
+           /* Workaround https://reviews.llvm.org/D131064 . */
+#   elif __clang_major__ < 15
+#       define __ARM_FEATURE_CRYPTO 1
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:178]: That workaround a bug of clang. It should not appear in this document. I just copy it from my test code.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+#   endif
+#   pragma clang attribute push (__attribute__((target("crypto"))), apply_to=function)
+#   define MBEDTLS_POP_TARGET_PRAGMA
+#elif defined(__GNUC__)
+#   if __GNUC__ < 6
+#       error "A more recent GCC is required for MBEDTLS_AES_C"
+#   else
+#       pragma GCC target ("arch=armv8-a+crypto")
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:186]: With `runtime detection`, I don't think so. 

This piece should be put into hardware acceleration module. In build scripts, if it is not supported, the module should be disabled. 

`#       error "A more recent GCC is required for MBEDTLS_AES_C"` should be changed to `#       error "A more recent GCC is required for MBEDTLS_AESCE_C"` . That might be more clearly

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+  - If possible, enable/disable hardware accelerations base on `CFLAGS`.
+
+NOTE: The algorithm can be implemented with different instruction sets on same architecture. For example, AES can be implemented with crypto extention or SVEAES extension in Arm64.
+
+
+## Suggestion
+
+### check config
+
+`check_config.h` MUST not include any archeticture special checks. This gives up the advantage of the current code. But that makes it architecture independent.
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:39]: >On a related note, `check_config.h` must not check compiler options.

Do you mean it has been included in other document? I did not get it when I wrote it.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
+# Build-time and runtime detection for hardware acceleration with advance CPU features
+
+This document describes how to support the algorithms that can do hardware acceleration with advance CPU features.
+
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:4]: I do not know that.  I will take a look. 
Where is document about Mbed TLS 4.0?

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+`check_config.h` MUST not include any archeticture special checks. This gives up the advantage of the current code. But that makes it architecture independent.
+
+For `build-time detection only`, we can enable the the checks in hardware acceleration module.
+
+There are a potential issue for `build-time detection only`, if library is built with some CPU modifer flags, how application known the parameter? Is it important ?
+
+### Configure options
+
+- `MBEDTLS_RUNTIME_DETECTION`: Enable runtime detection feature.
+- Target platform macros: `MBEDTLS_ARCH_IS_{X86,AMD64,ARM64}`
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:48]: That's for host build scripts. HOST CC should override it. 
Without it, the configuration options might be different between HOST CC and target CC for cross build.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+`check_config.h` MUST not include any archeticture special checks. This gives up the advantage of the current code. But that makes it architecture independent.
+
+For `build-time detection only`, we can enable the the checks in hardware acceleration module.
+
+There are a potential issue for `build-time detection only`, if library is built with some CPU modifer flags, how application known the parameter? Is it important ?
+
+### Configure options
+
+- `MBEDTLS_RUNTIME_DETECTION`: Enable runtime detection feature.
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:47]: > Should this really be a configuration option? 

Yes. But I think it should be renamed to `MBEDTLS_RUNTIME_C` now.

```c
static int aes_internal_setkey_enc(unsigned char *rk,
                                   const unsigned char *key,
                                   size_t bits)
{

#if defined(MBEDTLS_ARCH_IS_ARM64)
#if defined(MBEDTLS_AESCE_C) 
    if (mbedtls_aesce_has_support()) {
        return mbedtls_aesce_setkey_enc(rk, key, bits);
    }
#endif
#endif /* MBEDTLS_ARCH_IS_ARM64 */

#if !defined(MBEDTLS_AES_DISABLE_NO_ACCEL)
    return aes_setkey_enc_no_accel((uint32_t *) rk, key, bits);
#else
    return 0;
#endif

}
```

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+This module depends on `MBEDTLS_RUNTIME_DETECTION`. It provides cpu feature detection function and feature masks.
+
+files:
+- `library/runtime.h`
+- `library/runtime.c`
+
+I prefer implement the function as util macro in `runtime.h` like below.
+
+```c
+#if defined(MBEDTLS_RUNTIME_DETECTION)
```
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:101]: Sure. I will add it as a sub option of `MBEDTLS_RUNTIME_C`

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:48]: And there are many code to detect the architecture, I think we can replace them with our own macro

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:39]: `check compiler options` can reduce the risk of runtime failure. But I do not think it is a big problem.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:4]: >If a driver has fallback disabled, the subsequent drivers in the list are ignored for this algorithm.

Do the subsequent drivers exist in the target binary? 

How to reduce code size with one driver per algorithm in PSA drivers? This case is for known SOC.

I did not catch above points in [PSA drivers](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/proposed/psa-driver-interface.md)

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:178]: If the URL above is supposed to explain (I wasn't sure: it looks like it's related to the `#error` directive), I don't understand it. `__ARM_FEATURE_CRYPTO` doesn't appear on the linked page.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:186]: On second thoughts, this may be the right thing. We need to tell the compiler that the accelerated AES code, and _only the accelerated AES code_, can use the crypto instructions. So this needs to be only enabled around the accelerated AES code, and not in a header file.

When I first read this section, I thought this snippet was supposed to be in a header file. Reading this again, I think you might have intended it to be in the C file? Please clarify.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:39]: I don't think we've ever written a formal document to say “`check_config.h` must not check compiler options”, because that wasn't the purpose of `check_config.h`. Its intent is to check preprocessor definitions.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:4]: You select which implementations to include by passing a set of JSON files to `generate_driver_wrappers.py`. If you don't want to include a driver, don't pass its JSON file. Furthermore, if there are multiple drivers for an algorithm, only the first one is included (or more precisely: the build only includes the drivers up to the first one that has fallback=false).

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:47]: What I don't understand is why it should be a configuration option. Each configuration option increases the cost of testing. Why can't we always have `mbedtls_cpu_has_feature`, and it's a `static inline` constant function (so no cost in code size) if nothing needs the information?

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:186]: Sure. I will clarify it.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:47]: > it's a `static inline` constant function (so no cost in code size) if nothing needs the information?

It has cost in code size. The differences are the code size on different architecture.  With `MBEDTLS_RUNTIME_C`, we can just enable it with `#define mbedtls_cpu_has_feature(...) 1`,  the code size from `runtime detection` will be reduced. 

For x86/x64, the cost in code size is very small. from the [code size comparision](https://godbolt.org/z/xG4v9sYc9) , 14 extra instructions was generated. 

For Arm64/32, the issue is very complex. `sha512.c` shows the solutions. It is not small cost operation.
- With elf hwcap ( linux), the cost is small than x86/x64. 
- Apple/Windows, I guess it will be trapped into kernel mode. It is heavy cost for runtime function. It needs a profile function for it. It is a big cost in the code size.
- For others, we can profile the features 
  - profile with SIG_ILL signal
  - profile with reading sys registers if it is allowed.

**Comment by yuhaoth** [comment]
Not all comments is addressed.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (yanesca) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:4]: > Do we want the legacy mechanism to last for a long time? Or are we going to remove it in Mbed TLS 4.0?

We want to retire the legacy mechanism as soon as possible, however we only should do it when
1. the PSA Crypto Driver interface can do the same job
2. all use cases in TLS and X509 can work in PSA Crypto mode

I am not completely up to date on the progress in these areas, but I think we are pretty close to achieving 2. (if not already there), but there is some more work to be done to achieve 1. If we can get there by 4.0, I think we should retire the legacy mechanism in that release.

> How will the legacy mechanism work for PSA-only users?

Since they won't coexist for very long, I think we should go for the option that is most convenient for us and presents an at least marginally sensible interaction between the two mechanisms.

> The single feature detection function in this proposal makes more sense for core CPU features.

I am not sure about that, I always thought that these crypto specific extensions that are targeted at very specific algorithms/modules (like AES and SHA for example) would eventually be moved into a standard PSA Crypto Driver that happens to be supplied with the library. (Although in this case we still can have a single feature detection function, which would be called from the drivers.)

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (yanesca) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:130]: Why do we need a custom detection function? What would be the use case?

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (yanesca) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: I don't think this config option is needed either. In the corresponding WIP PR there are macros to detect whether the runtime config option is needed and error out if it is not present when needed or present when not needed:
https://github.com/Mbed-TLS/mbedtls/pull/7078/commits/32ac20ac047a91d8442b169036d68622bb3cf7b9#diff-4fc3811c7c53e1746e1056ac38a6462ad2160077a59898bbd2bc31f5938ac185R1101-R1116

How about instead of erroring out just turning `MBEDTLS_RUNTIME_C` on and off as needed and not having it as a config option in `config.h` at all?

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:130]: Now, we only support arm64/32 x86/x64. If user want to support risc-v , he can provide it.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: >Why can't we always have mbedtls_cpu_has_feature, and it's a static inline constant function (so no cost in code size) if nothing needs the information?

@gilles-peskine-arm , do you mean implement like bellow
```c
static inline int mbedtls_cpu_has_features(uint32_t flags)
{
#if defined(MBEDTLS_RUNTIME_C)
        return mbedtls_cpu_has_features_arm64(flags);
#else
       return (MBEDTLS_TARGET_CPU_FEATURES&flags)==flags;
#endif
}
```
I think I misunderstand your meaning.

[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: I just add a [table](https://github.com/Mbed-TLS/mbedtls/pull/7004/commits/c5324f3ed29906f165db15c726f2e6d908756558) for macros.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:4]: Should I abandon this proposal ? That really needs more efforts to implement it

But I will add `MBEDTLS_RUNTIME_C` and I think it can be reused in PSA drivers.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:68]: it is removed

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:215]: it is removed

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (yanesca) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:4]: > Should I abandon this proposal ? That really needs more efforts to implement it

I don't think we should abandon this proposal. What would need more effort to implement?

[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:4]: We should change sha256/512 and aes for x86 also. 

I just take a look at those module again. 
- AES for x86/x64, it implemented with assembly, that is simple, just replace runtime detection function.
- SHA512/256,  runtime detection function can be replace firstly. I have no idea about effort on other part.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:207]: It is removed.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:4]: I have update #7078 to remove many configuration options.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: I have update #7078 to remove many configuration options.   I guess it is suggestion from @gilles-peskine-arm  now.

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:130]: Arguably Risc-V is out of scope since we don't provide any Risc-V acceleration. If we ever add Risc-V acceleration, it should come with Risc-V feature detection.

But we need custom feature detection for arm! Unfortunately, on arm, feature detection requires the OS's collaboration. The instructions to query the CPU features rely on the OS. You can try an instruction and see if it faults, but catching the feault again relies on the OS. So we can provide methods for common OSes, but embedded OSes need to be able to provide their own.

[the remark this answers is not in the document]
Comment (yanesca) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:130]: Oh I see, thank you both

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: No, I mean something like:
```
static inline int mbedtls_cpu_has_features(uint32_t flags) {
#if defined(MBEDTLS_AES_USE_A64_CRYPTO_IF_PRESENT) || defined(MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT) || defined(MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT)
    return mbedtls_cpu_has_features_arm64(flags);
#else
    return 0;
#endif
}
```
I don't see a need for a separate compile-time option.

[the remark this answers is not in the document]
Comment (yanesca) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: I agree that we don't need `MBEDTLS_RUNTIME_C` in as a user facing high level compile time option in `mbedtls_config.h`.

We can have `MBEDTLS_RUNTIME_C` macro defined somewhere else (eg `build_info.h` or `runtime.h`) as a helper macro to improve readability and for convenience:

```C
#if defined(MBEDTLS_AES_USE_A64_CRYPTO_IF_PRESENT)    ||  \
    defined(MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT) ||  \
    defined(MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT)
#define MBEDTLS_RUNTIME_C
#endif
}
```

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:68]: I moved some cflags checks to source file in #6932 . Please take a look

File: docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: Understand. 

And your example code is wrong.  It should be
```c
#if defined(MBEDTLS_AES_CE_C)    ||  \
    defined(MBEDTL_AES_SVE_C) ||  \
    defined(MBEDTLS_AES_PURE_SOFTWARE_C)
#define MBEDTLS_RUNTIME_C
#endif
```

If any algorithm enables two accelerator , we should enable runtime detection.

[the remark this answers is not in the document]
Comment (yanesca) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: Indeed, my example code was still using the old 3 state macros, in the new proposal there wouldn't be any `MBEDTLS_AES_USE_A64_CRYPTO_IF_PRESENT` macros at all.

[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: I mean, if an algorithm has more than 1 accelerators , `MBEDTLS_RUNTIME_C` should be enabed. Not base on different algorithms. And I update my example
```c
#if MBEDTLS_AES_USE_A64_CRYPO + MBEDTLS_AES_USE_A64_SVE + MBEDTLS_AES_USE_PURE_SOFTWARE_C > 1
#define MBEDTLS_RUNTIME_C
#elif MBEDTLS_AES_USE_A64_CRYPO + MBEDTLS_AES_USE_A64_SVE + MBEDTLS_AES_USE_PURE_SOFTWARE_C == 1
/* Runtime detection was disbaled */
#else
#error "MBEDTLS_AES_C defined, But no accelerator provided"
#endif

#if MBEDTLS_SHA256_USE_A64_CRYPO + MBEDTLS_SHA256_USE_A64_SVE + MBEDTLS_SHA256_USE_PURE_SOFTWARE_C > 1
#define MBEDTLS_RUNTIME_C
#elif MBEDTLS_SHA256_USE_A64_CRYPO + MBEDTLS_SHA256_USE_A64_SVE + MBEDTLS_SHA256_USE_PURE_SOFTWARE_C == 1
/* Runtime detection was disbaled */
#else
#error "MBEDTLS_SHA256_C defined, But no accelerator provided"
#endif
```

[the remark this answers is not in the document]
Comment (yuhaoth) [reply docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md:76]: Done in #7078. Please double check if it is expected.
